File: 2015/php/6/main-part-1.php
<?php

function get_action($line){
    $line = explode(" ", $line);
    if ($line[0] == "turn"){
        $action = $line[1];
        $start = explode(",", $line[2]);
        $end = explode(",", $line[4]);
        return array($action, $start, $end);
    } else {
        $action = $line[0];
        $start = explode(",", $line[1]);
        $end = explode(",", $line[3]);
        return array($action, $start, $end);
    }
}

function main() {
    $argv = $_SERVER['argv'];
    $argc = $_SERVER['argc'];

    if ($argc < 2) {
        echo "Usage: php main-part1.php <input.txt>\n";
        exit(1);
    }

    $inputFile = $argv[1];
    $input = file_get_contents($inputFile);
    if ($input === false) {
        echo "Failed to read input file: $inputFile\n";
        exit(1);
    }

    $lines = explode("\n", $input);

    $grid = array_fill(0, 1000, array_fill(0, 1000, FALSE));

    foreach ($lines as $line) {
        if (empty($line)) {
            continue;
        }

        list($action, $start, $end) = get_action($line);
    

        for ($i = $start[0]; $i <= $end[0]; $i++) {
            for ($j = $start[1]; $j <= $end[1]; $j++) {
                if ($action == "on") {
                    $grid[$i][$j] = TRUE;
                } elseif ($action == "off") {
                    $grid[$i][$j] = FALSE;
                } elseif ($action == "toggle") {
                    $grid[$i][$j] = !$grid[$i][$j];
                }
            }
        }        
    }

    $count = 0;
    foreach ($grid as $row) {
        foreach ($row as $cell) {
            if ($cell) {
                $count++;
            }
        }
    }

    echo "Part 1: $count\n";
}

main();

File: 2015/haskell/6/main-part-1.hs
import System.Environment (getArgs)

main :: IO ()
main = do
  args <- getArgs
  case args of
    [filename] -> processFile filename
    _ -> putStrLn "Usage: main <input_file>"

processFile :: FilePath -> IO ()
processFile filename = do
  contents <- readFile filename
  let ls = lines contents
  let commands = map parseLine ls
  let grid' = foldl applyCommand grid commands
  let count = countLights grid'
  putStrLn $ "Part 1: " ++ show count

data Command = TurnOn Int Int Int Int | TurnOff Int Int Int Int | Toggle Int Int Int Int
  deriving (Show)

parseLine :: String -> Command
parseLine s = case words s of
  ["turn", "on", coord1, "through", coord2] -> parseCommand coord1 coord2 TurnOn
  ["turn", "off", coord1, "through", coord2] -> parseCommand coord1 coord2 TurnOff
  ["toggle", coord1, "through", coord2] -> parseCommand coord1 coord2 Toggle
  _ -> error $ "Invalid command: " ++ s

parseCommand :: String -> String -> (Int -> Int -> Int -> Int -> Command) -> Command
parseCommand coord1 coord2 f =
  let [x1, y1] = map read $ split ',' coord1
      [x2, y2] = map read $ split ',' coord2
   in f x1 y1 x2 y2

split :: (Eq a) => a -> [a] -> [[a]]
split _ [] = []
split delim xs =
  let (before, after) = span (/= delim) xs
   in before : split delim (drop 1 after)

type Grid = [[Bool]]

grid :: Grid
grid = replicate 1000 (replicate 1000 False)

applyCommand :: Grid -> Command -> Grid
applyCommand g (TurnOn x1 y1 x2 y2) = updateGrid g x1 y1 x2 y2 (const True)
applyCommand g (TurnOff x1 y1 x2 y2) = updateGrid g x1 y1 x2 y2 (const False)
applyCommand g (Toggle x1 y1 x2 y2) = updateGrid g x1 y1 x2 y2 not

updateGrid :: Grid -> Int -> Int -> Int -> Int -> (Bool -> Bool) -> Grid
updateGrid g x1 y1 x2 y2 f =
  let rows = take (y2 - y1 + 1) $ drop y1 g
      rows' = map (updateRow x1 x2 f) rows
   in take y1 g ++ rows' ++ drop (y2 + 1) g

updateRow :: Int -> Int -> (Bool -> Bool) -> [Bool] -> [Bool]
updateRow x1 x2 f row =
  let (left, right) = splitAt x1 row
      (middle, right') = splitAt (x2 - x1 + 1) right
   in left ++ map f middle ++ right'

countLights :: Grid -> Int
countLights g = length $ concatMap (filter id) g